ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java

```diff
@@ -1164,13 +1195,49 @@ public Node visitColumnReference(SqlBaseParser.ColumnReferenceContext context) {
         throw new InvalidColumnReferenceException("Field " + columnName + " is ambiguous.");
       }
     } else {
-      Expression baseExpression =
-          new QualifiedNameReference(
-              getLocation(context),
-              QualifiedName.of(dataSourceExtractor.getFromAlias())
-          );
-      return new DereferenceExpression(getLocation(context), baseExpression, columnName);
+      if (isAmbiguousColumnName(
+          columnName,
+          suffixFieldName,
+          dataSourceExtractor.getFromName(),
+          dataSourceExtractor.getFromAlias(),
+          dataSourceExtractor.getFromSchema())) {
+        throw new KsqlException(String.format("Ambiguous column name %s. You have stream/table"
+            + " with the same name/alias. Use stream/table name or"
+            + " alias as prefix when there is a column with the same name.", columnName));
+      }
+      if (columnName.equalsIgnoreCase(dataSourceExtractor.getFromAlias())
+          || columnName.equalsIgnoreCase(dataSourceExtractor.getFromName())) {
+        return new QualifiedNameReference(
+            getLocation(context),
+            QualifiedName.of(columnName)
+        );
+      } else {
+        Expression baseExpression =
+            new QualifiedNameReference(
+                getLocation(context),
+                QualifiedName.of(dataSourceExtractor.getFromAlias())
+            );
+        return new DereferenceExpression(getLocation(context), baseExpression, columnName);
+      }
+    }
+  }
+
+  private boolean isAmbiguousColumnName(
+      final String columnName,
+      final String suffixName,
+      final String sourceName,
+      final String sourceAlias,
+      final Schema sourceSchema
+  ) {
+    boolean isASchemaColumn = sourceSchema.fields().stream()
+        .anyMatch(field -> columnName.equalsIgnoreCase(field.name()));
+    if (suffixName == null) {
+      return isASchemaColumn
+          && (columnName.equalsIgnoreCase(sourceName) || columnName.equalsIgnoreCase(sourceAlias));
     }
+    return isASchemaColumn
+        && !columnName.equalsIgnoreCase(suffixName)
```

Comment:
Yes, that's the intended behavior and as you said we should document it an keep it simple. In the above case for instance `S.S` will assume that the first `S` is the stream name and the second `S` is the column name.

Reply:
Let me just repeat that back so I know I've understood that correctly:

we're saying that if the first part of the column name matches the stream/table name/alias, then we assume that's what it is?

e.g. 
```
CREATE STREAM S (S STRUCT<S INTEGER, C VARCHAR>) ...;
SELECT S.S FROM S;   -- selects whole struct.
SELECT S.C FROM S;   -- fails as there is no first-level column C.
```

I think there is still room for people to grok the first statement wrong. I think most people would expect it to select the integer.

Can we not just fail if the table/stream name/alias matches any first-level column identifier?

Reply:
@big-andy-coates do you mean fail the ddl statement? Is the reasoning that this is an easier error to deal with then dealing with unexpected successful behavior later on? 

I think if a user wants a stream and one of the columns to have the same name we should let them. If a later column reference has an ambiguity we should have a sane way of resolving it. Detecting if there actually is an ambiguity is a bit complex (especially if you have subscripts). Defaulting to the stream name whenever the first level of the column reference matches the stream name seems ok for now.

Reply:
Hummm... what I'm worried about is allowing syntax that is very easily misinterpreted by the next person that comes along and reads it.  Given:

```
CREATE STREAM S (S STRUCT<S INTEGER, C VARCHAR>) ...;
SELECT S.S FROM S;   -- selects whole struct.
```

Someone coming along with SQL experience, but without in-depth knowledge of the nuances of KSQL, would likely think its selecting the integer field. 

Just documenting "Well, we treat it this way" is all well and good, but it requires users to have read the documentation. And as we all know, not many do!

Hence wondering if just failing on the DLL statement isn't the way to go. With a message like:

> STRUCT column name can not match Stream name or alias. Rename the stream or column. You can use 'CREATE STREAM S_ WITH(KAFKA_TOPIC="S") ...' to keep the underlying topic name the same.

If we later think of a better way to handle this, then we can always allow such statements later. But if we accept them now, its harder to stop later.

Reply:
I agree with @big-andy-coates that this is confusing.

> Can we not just fail if the table/stream name/alias matches any first-level column identifier?

This is one option to sort this out -- it's IMHO both simple plus easy to understand and remember (for the user ) and to explain (for us/KSQL error messages) --.  And I'd rather have this option than having to rely on users reading the docs (99% of the time they won't).

Open to other suggestions, of course, but I'd very much go with Andy's suggestion rather than with guessing what field would or would not be selected.

Reply:
@big-andy-coates , @miguno Enforcing the requirement in DDL statemens to not have column name the same as stream name won't solve the problem. Users still cam use alias in their queries the same as column name and we will have to resolve this in query. Also in case of JOIN you can have column name from one side of the join be the same as the stream/table name from the other side and again we will  have ambiguity. 
I think the current behavior, defaulting to assume the stream name is the prefix is preferable.

Reply:
Ah yes, it's more complex than I thought - thanks @hjafarpour.

The tricky one is the join. For join's you're joining two, or more, definitions together. I think its fine to fail a single command because its got ambiguities in it, but it's harder to fail something because the combination of multiple commands is leading to ambiguities. 

I don't _like_ the idea of not allowing certain stream / table names. But, without a better a approach, I think its the lesser of two evils.

```
CREATE STREAM X (Y STRUCT<Y INTEGER>, X INTEGER) ...;  -- valid
CREATE STREAM Y (X STRUCT<X INTEGER>, Y INTEGER) ...;  -- valid

SELECT X.X, Y.Y FROM X JOIN Y ON ...  -- very ambiguous!
```

One option, similar to what you're proposing, is to say that a struct path must always be fully qualified, i.e. it must always start with the table/stream name or alias.   This would avoid the ambiguity. The above would therefore become:

```
SELECT X.X, Y.Y FROM X JOIN Y ON ...  -- fails, as no suitable path can be found.

SELECT Y.X, Y.X.X, X.Y, X.Y,Y FROM X JOIN Y ON ... - explicit and valid.
```

I think its ok to say the path into a struct must always be fully qualified, as it avoids ambiguity.

The other alternative is to only fail where it is ambiguous.  So,

```
CREATE STREAM X (X STRUCT<X INTEGER>, Y STRUCT<Y INTEGER>) ...;  -- valid

SELECT Y.Y from X; -- is fine as its not ambiguous.
SELECT X.X.X from X; -- full qualified, so also fine.
SELECT X.X from X; -- fails as its ambiguous.
```

This is probably the nicest for the user. Though, if we support updating the upstream schemas we're in trouble. Consider the future, where KSQL doesn't require CS/CT statements, as it can just _discover_ streams and tables and their schemas in the cluster. Now we're in a situation where the schema of a source can evolve over time. This opens up the possibility of statements that once were unambiguous becoming ambiguous.

So... I think the safest option is to _require_ fully-qualified struct paths.

What say y'all?

Reply:
I'm a +1 on using fully-qualified paths

Reply:
I think that might be the same as I was saying... maybe. Yes, 'X.X' would select the whole struct. But also, importantly, structs _always_ require a fully qualified path, even where it wouldn't be ambiguous to drop the table/stream name/alias e.g.

```
CREATE STREAM Bob (X STRUCT<X INTEGER>, Y STRUCT<Y INTEGER>) ...;
SELECT X.X FROM Bob; -- fails: Struct paths must be fully qualified.
```

The _benefit_ of this approach is that users will quickly come to _know_ that struct paths are always fully qualified. Therefore, they are much less likely to miss read something like:

```
CREATE STREAM X (X STRUCT<X INTEGER>) ...;
SELECT X.X FROM X;   -- selects the whole struct.
```

If we only require paths to be fully qualified when they are ambiguous users don't learn these nuances.

Reply:
Interpreting the source name as referring to the source is more intuitive to me. The concept of resolving ambiguities with some precedence protocol should be familiar to people using computers.

That said, I'm +1 for requiring fully qualified paths for struct dereferences for the following reasons:

Like @big-andy-coates  said - requiring fully qualified paths for structs will trip users up more. This is both good and bad. A little bit bad because its not the most pleasant experience as you're getting started, but mostly good because with a decent error message its easy to learn and avoids confusion in the corner case where a column has the same name as the source.

Always interpreting the source name as referring to the source may break queries that work today, which we need to start taking seriously.

> if we support updating the upstream schemas we're in trouble. Consider the future, where KSQL doesn't require CS/CT statements, as it can just discover streams and tables and their schemas in the cluster. Now we're in a situation where the schema of a source can evolve over time. This opens up the possibility of statements that once were unambiguous becoming ambiguous.

This is an interesting point. I think for most other changes the schema-registry's compatibility checks can ensure that schemas don't change in an incompatible way. But they don't help for the join case.

Reply:
I think here we are optimizing for a 1% case by imposing unnecessary requirement on 99% of the cases. We are also deviating from common practice of optional stream/table name/alias prefixing for columns. The only case we need to address ambiguity is when we have a situation like the following:

`CREATE STREAM A (A STRUCT<A INTEGER>, B STRUCT<B STRING>) WITH ...`

In the above case we have a stream, column and field of the column all having the same name. In every other case we won't have any ambiguity. The following behavior is intuitive and also does not require mandatory prefix for struct columns:

`SELECT A FROM A;         -- will return struct of column A`
`SELECT A.A FROM A;      -- will return struct of column A`
`SELECT A.A.A FROM A;   -- will return integer field with name A of the struct column A`
`SELECT B FROM A;           -- no ambiguity and no need for prefixing with table name`
`SELECT B.B FROM A;     -- again, no ambiguity and no need to enforce having table name prefix`

Enforcing users to always have stream/table name/alias as prefix for any struct type regardless of having any ambiguity would not create a positive user experience and I can imagine how many times users will be frustrated when getting error for not having the prefix.
I think we should also ask @rmoff and @blueedgenick 's opinion since they are more fluent in SQL and know our target users better.

Reply:
One concern with requiring full de-references for struct types is that you could run into potentially major inconveniences when the stream/table names are very long. For example, If you have a table named named `organization_activity_by_day` (a real example), then this can get pretty long pretty fast, especially if the struct field is verbose (and more so if individual column names are long).

That said, resolving the ambiguity problem using precedence rules is adds more cognitive overhead to the user. Similar looking statements would start meaning something different based on identifiers involved:

```
CREATE STREAM FOO (BAR STRUCT<X INTEGER,  ..>) WITH ..;
CREATE STREAM BAR (X bigint, Y bigint) WITH ...;

SELECT BAR.X, X FROM FOO JOIN BAR ON ...;
```

From what I have understood, this would select the same `X` (from `BAR`) twice if we use the suggested precedence rules, and is very non obvious. I think requiring struct types to be fully qualified will enable users to educate themselves about what to expect, albeit at the cost of some inconvenience.

On the balance, I think some inconvenience for a less loaded mental model is preferable, and am hence leaning towards requiring struct fields to be fully qualified.

Reply:
@apurvam the example you have exactly shows that regardless of requiring or not requiring fully qualified path user needs to make the assumption that the stream/table name/alias has the precedence. If you enforce prefix on struct fields your will have the exact same query. With this assumption you can easily remove the ambiguity. 
So in both cases, the above query will return the same results.

Now let's assume user wants to select field `X` in the struct column `BAR` from stream `FOO` and also select the integer field `X` from stream `B`. For both mandatory and non mandatory prefix the queries will be the same:

``` SELECT FOO.BAR.X, X FROM FOO JOIN BAR ON ...;```

The main argument is that the query results for mandating and not mandating stream/table name/alias for struct fields is the same and mandating the prefix only adds extra burden  on user to alway add prefix even when it's not needed.

Reply:
Thanks for your feedback @rmoff . The problem is that even when we have ambiguity we won't be able show any error message. Consider the following example:

```CREATE STREAM A (A STRUCT<A INTEGER>, B STRUCT<B STRING>) WITH ...```
```SELECT A.A FROM A; --regardless of enforcing full path or not this won't fail and the result will be the same```

In the above example we will run the query and select the whole struct column and regardless of enforcing the full path or not it will always select the struct column instead of struct field. 
As for other SQL systems most of the traditional ones as you mentioned may not support nested types but most of the recent ones that are being used in big data community and streaming such as Hive, Impala, SparkSQL, and Calcite based systems (Flink, Phoenix, Kylin,...) do use dot notation and don't mandate the table name/alias prefixing for struct columns.

Reply:
adding another couple cents to the growing pile of pocket change: i happen to agree in large part with both @rmoff and @hjafarpour on this one :-) ..while it's true that most of the traditional RDBMS don't make this kind of addressing of nested structures very easy, with most of them resorting to function-based access, it's a matter of persepctive: they are all so old they never thought of having structured data as a type until much later, when it was hard to retrofit. So internally they mostly think of something like json as a varchar or clob that you can force "special rules" onto by use of a family of json-parsing functions. Similarly, in the past we thought that normalization and the assignment of attributes to tables was enough to handle any kind of "data structure" so `STRUCT` as a type made no sense. In the era of big-data systems it became much more common to allow access to json-like structures and `STRUCT`s natively, with a couple of prominent systems like Hive doing the same thing for avro data too. 
So, the questions of user expectations and familiarity might largely revolve around what you're used to, and i'd strongly favor going for the "big-data mindset" on this one, as it's closer to how we think in terms of native support for complex datatypes. 
I've had success in the past with avro data in Hive, where avro data translates neatly into a defined structure of nested maps, arrays, structs, and primitves, all of which have well-defined syntax for accessing them (with `[]`, `#` and `:` characters for array, map, and struct entries respectively). To query JSON in Hive, IIRC (worth double-checking this one, might have been improved upon!) you have to pre-declare a strucutre of maps/arrays/structs in the table schema, and then access as i just described for avro. So the query-writing becomes easy, and the question of dot-notation ambiguity never arises because of the use of different characters, but it often feels super clumsy to have to schematize your JSON this way.
Which brings us to Drill... not my favorite but it's focus on usability makes it possible to write dot-path access to nested fields within a query without first declaring any schema, which might be the closest to what we are looking for here ?
In summary, if i had to vote for one option, i'd probably go with something close to how Hive et al do this, for the largest community of folks who are already familiar with such an option, and therefore introduce the notion of different separator characters for different purposes. This works well for avro but note the trade-offs when querying JSON with this approach - it can still be done without declaring the structure ahead of time but the user needs to think of their json explicitly as a series of maps and arrays and then query it as such. I doubt that seems too foreign to any programmer types in the intended audience, the only question might be how intuitive it is to a more sql-background user.